python/instrumentation/openinference-instrumentation-llama-index/tests/openinference/instrumentation/llama_index/test_handler.py

```diff
@@ -0,0 +1,521 @@
+import asyncio
+import json
+import logging
+import random
+from collections import defaultdict
+from concurrent.futures import ThreadPoolExecutor
+from contextlib import suppress
+from itertools import count
+from typing import (
+    Any,
+    AsyncIterator,
+    DefaultDict,
+    Dict,
+    Generator,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    Tuple,
+    cast,
+)
+
+import openai
+import pytest
+from httpx import AsyncByteStream, Response, SyncByteStream
+from llama_index.core import Document, ListIndex, Settings
+from llama_index.core.base.response.schema import StreamingResponse
+from llama_index.core.callbacks import CallbackManager
+from llama_index.core.schema import TextNode
+from llama_index.llms.openai import OpenAI  # type: ignore
+from openinference.instrumentation import using_attributes
+from openinference.instrumentation.llama_index import LlamaIndexInstrumentor, _legacy_llama_index
+from openinference.semconv.trace import (
+    DocumentAttributes,
+    EmbeddingAttributes,
+    MessageAttributes,
+    OpenInferenceMimeTypeValues,
+    OpenInferenceSpanKindValues,
+    SpanAttributes,
+    ToolCallAttributes,
+)
+from opentelemetry import trace as trace_api
+from opentelemetry.sdk import trace as trace_sdk
+from opentelemetry.sdk.trace import ReadableSpan
+from opentelemetry.sdk.trace.export import SimpleSpanProcessor
+from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter
+from respx import MockRouter
+
+for name, logger in logging.root.manager.loggerDict.items():
+    if name.startswith("openinference.") and isinstance(logger, logging.Logger):
+        logger.setLevel(logging.DEBUG)
+        logger.handlers.clear()
+        logger.addHandler(logging.StreamHandler())
+
+
+@pytest.mark.skipif(
+    _legacy_llama_index(),
+    reason="legacy callback system",
+)
+@pytest.mark.parametrize(
+    "is_stream,is_async",
+    [
+        (False, False),
+        (False, True),
+        (True, False),
+        # FIXME: stream + async is not supported by LlamaIndex as of v0.9.33
+        # (True, True),
+    ],
+)
+@pytest.mark.parametrize(
+    "status_code", [200]
+)  # 400 has been removed because retries have been added and can't be easily modified
+@pytest.mark.parametrize("use_context_attributes", [False, True])
+def test_handler_basic_retrieval(
+    is_async: bool,
+    is_stream: bool,
+    status_code: int,
+    use_context_attributes: bool,
+    respx_mock: MockRouter,
+    in_memory_span_exporter: InMemorySpanExporter,
+    nodes: List[Document],
+    chat_completion_mock_stream: Tuple[List[bytes], List[Dict[str, Any]]],
+    session_id: str,
+    user_id: str,
+    metadata: Dict[str, Any],
+    tags: List[str],
+) -> None:
+    n = 10  # number of concurrent queries
```

Comment:
Is the threadpool needed on this test case?

Reply:
I don't seem to have a problem with the callstack using the IDE if that's what you're alluding to.

<img width="374" alt="Screenshot 2024-05-16 at 11 06 29 AM" src="https://github.com/Arize-ai/openinference/assets/80478925/330c3f9b-40a0-4c06-ad0e-ae33c946f57e">

Reply:
I think I'm not clear what purpose the threadpool is serving in this test case. Could we check the same things with a single question running in the main thread?

Reply:
Since the instrumentor is a singleton, there's no a priori assurance that the it will handle concurrency. So this is a simulation as confirmation.